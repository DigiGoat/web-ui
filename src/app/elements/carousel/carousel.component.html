<div #carousel id="goatImageCarousel" class="carousel slide" test-id="carousel">
  <div class="carousel-indicators">
    <button *ngFor="let image of images; index as i" type="button" data-bs-target="#goatImageCarousel" [class.active]="i === 0"
      [attr.aria-current]="i === 0" attr.aria-label="Slide {{i}}" [attr.data-bs-slide-to]="i" test-id="image-tab"></button>
  </div>
  <div class="carousel-inner d-flex align-items-center color-scheme-secondary" style="max-height: 75vh;">
    <div *ngFor="let image of images; index as i" class="carousel-item" [class.active]="i === 0">
      <img [ngSrc]="image.file" [alt]="image.alt" width="640" height="400" class="object-fit-contain w-100"
        [priority]="i < 2 || i === images!.length-1" test-id="image">
    </div>
  </div>
  <button class="carousel-control-prev" type="button" data-bs-target="#goatImageCarousel" data-bs-slide="prev">
    <span class="carousel-control-prev-icon" aria-hidden="true" test-id="control-icons"></span>
    <span class="visually-hidden">Previous</span>
  </button>
  <button class="carousel-control-next" type="button" data-bs-target="#goatImageCarousel" data-bs-slide="next">
    <span class="carousel-control-next-icon" aria-hidden="true" test-id="control-icons"></span>
    <span class="visually-hidden">Next</span>
  </button>
</div>
