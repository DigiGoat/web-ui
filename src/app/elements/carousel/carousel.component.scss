.carousel-indicators>* {
  background-color: var(--bs-emphasis-color) !important;
}

[data-bs-theme=dark] .carousel-control-next-icon,
[data-bs-theme=dark] .carousel-control-prev-icon {
  filter: inherit !important; //color: var(--bs-emphasis-color) !important;
}

[data-bs-theme=light] .carousel-control-next-icon,
[data-bs-theme=light] .carousel-control-prev-icon {
  filter: invert(1) grayscale(100) !important; //color: var(--bs-emphasis-color) !important;
}