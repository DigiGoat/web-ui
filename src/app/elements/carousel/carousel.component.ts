import { Component, Input, ViewChild, ViewEncapsulation, type AfterViewInit, type ElementRef, type OnDestroy, type OnInit } from '@angular/core';
import { Meta } from '@angular/platform-browser';
import { ConfigService } from '../../services/config/config.service';
import type { ImageEntry } from '../../services/image/image.service';
import { PlatformService } from '../../services/platform/platform.service';

@Component({
  selector: 'app-carousel',
  templateUrl: './carousel.component.html',
  styleUrl: './carousel.component.scss',
  encapsulation: ViewEncapsulation.None,
})
export class CarouselComponent implements AfterViewInit, OnDestroy, OnInit {
  constructor(private platformService: PlatformService, private meta: Meta, private configService: ConfigService) { }
  ngOnInit() {
    for (const image of this.images) {
      this.meta.addTag({ name: 'og:image', content: this.configService.link ? new URL(image.file, this.configService.link).toString() : image.file });
      //this.meta.addTags([{ name: 'og:image:width', content: '400' }, { name: 'og:image:height', content: '200' }], true);
      if (image.alt) {
        this.meta.addTag({ name: 'og:image:alt', content: image.alt });
      }
    }
  }
  @Input() images: ImageEntry[] = [];

  ngOnDestroy() {
    this.carousel?.dispose();
  }

  @ViewChild('carousel') carouselElement!: ElementRef<HTMLDivElement>;
  private carousel?: bootstrap.Carousel;
  ngAfterViewInit() {
    if (this.platformService.isBrowser) {
      this.carousel = bootstrap.Carousel.getOrCreateInstance(this.carouselElement.nativeElement);
    }
  }
}
