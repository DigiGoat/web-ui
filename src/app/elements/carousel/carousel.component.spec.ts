import { ComponentFixture, TestBed } from '@angular/core/testing';

import type { Meta } from '@angular/platform-browser';
import type { ConfigService } from '../../services/config/config.service';
import type { PlatformService } from '../../services/platform/platform.service';
import { CarouselComponent } from './carousel.component';

const bootstrap = { Carousel: { getOrCreateInstance: jest.fn().mockReturnValue({ dispose: jest.fn() }) } };
describe('CarouselComponent', () => {
  let component: CarouselComponent;
  let fixture: ComponentFixture<CarouselComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    (window as unknown as { bootstrap: object; }).bootstrap = bootstrap;
    await TestBed.configureTestingModule({
      declarations: [CarouselComponent]
    })
      .compileComponents();
    fixture = TestBed.createComponent(CarouselComponent);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
    //component['configService'] = jest.fn().mockReturnValue(ConfigServiceMock) as unknown as ConfigService;
  });
  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });
  describe('On the Server', () => {
    beforeEach(() => {
      component['platformService'] = {
        isServer: true,
        isBrowser: false,
      } as PlatformService;
      fixture.detectChanges();
    });
    it('should not create the carousel', () => {
      expect(bootstrap.Carousel.getOrCreateInstance).not.toHaveBeenCalledTimes(1);
    });
  });
  describe('In a Browser', () => {
    beforeEach(() => {
      component['platformService'] = {
        isServer: false,
        isBrowser: true,
      } as PlatformService;
      fixture.detectChanges();
    });
    it('should create the carousel', () => {
      expect(bootstrap.Carousel.getOrCreateInstance).toHaveBeenCalledTimes(1);
    });
    it('should dispose the carousel when destroyed', () => {
      fixture.destroy();
      expect(component['carousel']!.dispose).toHaveBeenCalledTimes(1);
    });
  });
  describe('With Images', () => {
    const meta = { addTag: jest.fn() };
    beforeEach(() => {
      component['meta'] = meta as unknown as Meta;
      component['images'] = [{ file: 'TEST_IMAGE_1' }, { file: 'TEST_IMAGE_2', alt: 'TEST_ALT_1' }];
    });
    describe('With a configured link', () => {
      beforeEach(() => {
        component['configService'] = { link: 'https://test.link' } as ConfigService;
        fixture.detectChanges();
      });
      it('should add image meta tags', () => {
        expect(component['meta'].addTag).toHaveBeenCalledTimes(3);
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(1, { 'content': 'https://test.link/TEST_IMAGE_1', 'name': 'og:image' });
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(2, { 'content': 'https://test.link/TEST_IMAGE_2', 'name': 'og:image' });
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(3, { 'content': 'TEST_ALT_1', 'name': 'og:image:alt' });
      });
      it('should make the carousel', () => {
        const element = html.querySelectorAll('[test-id=carousel]');
        expect(element).toHaveLength(1);
      });
      it('should have an indicator for each image', () => {
        const element = html.querySelectorAll('[test-id=image-tab]');
        expect(element).toHaveLength(2);
      });
      it('should have images', () => {
        const element = html.querySelectorAll('[test-id=image]');
        expect(element).toHaveLength(2);
      });
      it('should have control icons', () => {
        const element = html.querySelectorAll('[test-id=control-icons]');
        expect(element).toHaveLength(2);
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
    describe('Without a configured link', () => {
      beforeEach(() => {
        fixture.detectChanges();
      });
      it('should add image meta tags', () => {
        expect(component['meta'].addTag).toHaveBeenCalledTimes(3);
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(1, { 'content': 'TEST_IMAGE_1', 'name': 'og:image' });
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(2, { 'content': 'TEST_IMAGE_2', 'name': 'og:image' });
        expect(component['meta'].addTag).toHaveBeenNthCalledWith(3, { 'content': 'TEST_ALT_1', 'name': 'og:image:alt' });
      });
      it('should make the carousel', () => {
        const element = html.querySelectorAll('[test-id=carousel]');
        expect(element).toHaveLength(1);
      });
      it('should have an indicator for each image', () => {
        const element = html.querySelectorAll('[test-id=image-tab]');
        expect(element).toHaveLength(2);
      });
      it('should have images', () => {
        const element = html.querySelectorAll('[test-id=image]');
        expect(element).toHaveLength(2);
      });
      it('should have control icons', () => {
        const element = html.querySelectorAll('[test-id=control-icons]');
        expect(element).toHaveLength(2);
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
  });
});
