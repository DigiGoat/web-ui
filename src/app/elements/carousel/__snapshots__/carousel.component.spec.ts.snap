// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselComponent With Images With a configured link should match the snapshot 1`] = `
<app-carousel
  carousel={[Function Object]}
  carouselElement={[Function _ElementRef]}
  configService={[Function Object]}
  images={[Function Array]}
  meta={[Function Object]}
  platformService={[Function PlatformService]}
>
  <div
    class="carousel slide"
    id="goatImageCarousel"
    test-id="carousel"
  >
    <div
      class="carousel-indicators"
    >
      <button
        aria-current="true"
        aria-label="Slide 0"
        class="active"
        data-bs-slide-to="0"
        data-bs-target="#goatImageCarousel"
        test-id="image-tab"
        type="button"
      />
      <button
        aria-current="false"
        aria-label="Slide 1"
        data-bs-slide-to="1"
        data-bs-target="#goatImageCarousel"
        test-id="image-tab"
        type="button"
      />
    </div>
    <div
      class="carousel-inner d-flex align-items-center color-scheme-secondary"
      style="max-height: 75vh;"
    >
      <div
        class="carousel-item active"
      >
        <img
          alt="undefined"
          class="object-fit-contain w-100"
          height="400"
          test-id="image"
          width="640"
        />
      </div>
      <div
        class="carousel-item"
      >
        <img
          alt="TEST_ALT_1"
          class="object-fit-contain w-100"
          height="400"
          test-id="image"
          width="640"
        />
      </div>
    </div>
    <button
      class="carousel-control-prev"
      data-bs-slide="prev"
      data-bs-target="#goatImageCarousel"
      type="button"
    >
      <span
        aria-hidden="true"
        class="carousel-control-prev-icon"
        test-id="control-icons"
      />
      <span
        class="visually-hidden"
      >
        Previous
      </span>
    </button>
    <button
      class="carousel-control-next"
      data-bs-slide="next"
      data-bs-target="#goatImageCarousel"
      type="button"
    >
      <span
        aria-hidden="true"
        class="carousel-control-next-icon"
        test-id="control-icons"
      />
      <span
        class="visually-hidden"
      >
        Next
      </span>
    </button>
  </div>
</app-carousel>
`;

exports[`CarouselComponent With Images Without a configured link should match the snapshot 1`] = `
<app-carousel
  carousel={[Function Object]}
  carouselElement={[Function _ElementRef]}
  configService={[Function ConfigService]}
  images={[Function Array]}
  meta={[Function Object]}
  platformService={[Function PlatformService]}
>
  <div
    class="carousel slide"
    id="goatImageCarousel"
    test-id="carousel"
  >
    <div
      class="carousel-indicators"
    >
      <button
        aria-current="true"
        aria-label="Slide 0"
        class="active"
        data-bs-slide-to="0"
        data-bs-target="#goatImageCarousel"
        test-id="image-tab"
        type="button"
      />
      <button
        aria-current="false"
        aria-label="Slide 1"
        data-bs-slide-to="1"
        data-bs-target="#goatImageCarousel"
        test-id="image-tab"
        type="button"
      />
    </div>
    <div
      class="carousel-inner d-flex align-items-center color-scheme-secondary"
      style="max-height: 75vh;"
    >
      <div
        class="carousel-item active"
      >
        <img
          alt="undefined"
          class="object-fit-contain w-100"
          height="400"
          test-id="image"
          width="640"
        />
      </div>
      <div
        class="carousel-item"
      >
        <img
          alt="TEST_ALT_1"
          class="object-fit-contain w-100"
          height="400"
          test-id="image"
          width="640"
        />
      </div>
    </div>
    <button
      class="carousel-control-prev"
      data-bs-slide="prev"
      data-bs-target="#goatImageCarousel"
      type="button"
    >
      <span
        aria-hidden="true"
        class="carousel-control-prev-icon"
        test-id="control-icons"
      />
      <span
        class="visually-hidden"
      >
        Previous
      </span>
    </button>
    <button
      class="carousel-control-next"
      data-bs-slide="next"
      data-bs-target="#goatImageCarousel"
      type="button"
    >
      <span
        aria-hidden="true"
        class="carousel-control-next-icon"
        test-id="control-icons"
      />
      <span
        class="visually-hidden"
      >
        Next
      </span>
    </button>
  </div>
</app-carousel>
`;
